### PR TITLE
[Spark] Make DeltaHistoryManager.getHistory aware of ICT

### PR DESCRIPTION
**Pull request by dhruvarya-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Return the ICT (In-Commit Timestamp) timestamp for ICT-enabled ranges in DeltaHistoryManager.getHistory when ICT is currently enabled.

## How was this patch tested?

New tests:
1. When all of history has ICT enabled.
2. When requested history is not an ICT range.
3. When requested range contains both ICT and non-ICT commits.

## Does this PR introduce _any_ user-facing changes?

No visible differences for the user, the timestamp returned by `DESCRIBE HISTORY` will be different depending on whether ICT is currently enabled.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 1]
```diff
@@ -115,8 +117,60 @@ class DeltaHistoryManager(
...
+  /**
+   * Get the commit information of the Delta table from commit `[start, end)`. If `end` is `None`,
+   * we return all commits from start to now.
+   * @param start The start of the commit range, inclusive.
+   * @param end The end of the commit range, exclusive.
+   */
+  def getHistory(
+      start: Long,
+      end: Option[Long] = None,
+      snapshotOpt: Option[Snapshot] = None): Seq[DeltaHistory] = {
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:137]: Remove `snapshotOpt` and rely on unsafeVolatileSnapshot instead. That will simplify the interface without any performance regressions.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaHistoryManagerSuite.scala [hunk 2]
```diff
@@ -628,12 +628,12 @@ abstract class DeltaHistoryManagerBase extends DeltaTimeTravelTests
         assert(filterUsageRecords(usageRecords, "deltaLog.update").size === expectedLogUpdates)
       }
 
-      testGetHistory(start = 0, endOpt = Some(2), versions = Seq(2, 1, 0), expectedLogUpdates = 0)
-      testGetHistory(start = 1, endOpt = Some(1), versions = Seq(1), expectedLogUpdates = 0)
+      testGetHistory(start = 0, endOpt = Some(3), versions = Seq(2, 1, 0), expectedLogUpdates = 0)
```
Comment (sumeet-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaHistoryManagerSuite.scala:631]: Undo these changes as well.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 3]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
    * chronological order.
    */
   def getHistory(limitOpt: Option[Int]): Seq[DeltaHistory] = {
+    val snapshot = deltaLog.update()
     val listStart = limitOpt.map { limit =>
-      math.max(deltaLog.update().version - limit + 1, 0)
+      math.max(snapshot.version - limit + 1, 0)
     }.getOrElse(getEarliestDeltaFile(deltaLog))
-    getHistory(listStart)
+    getHistory(listStart, snapshotOpt = Some(snapshot))
```
[the remark this answers is not in the document]
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:74]: That's a good idea, it will also keep the function signature stable. Done.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 4]
```diff
@@ -115,8 +117,60 @@ class DeltaHistoryManager(
...
+    val monotonizedCommits = if (isRangeICT) {
+      info.collect()
+    } else {
+      monotonizeCommitTimestamps(info.collect())
+    }
+    monotonizedCommits.reverse
+  }
+
+  /**
+   * Get the commit information of the Delta table from commit `[start, end)`. If `end` is `None`,
```
[the remark this answers is not in the document]
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:129]: Done.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 5]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
...
-   * we return all commits from start to now.
+   * Get the commit information of the Delta table from commit `[start, end)`.
+   *
+   * @param isRangeICT Whether the commit range is ICT-enabled.
+   *                   If `true`, all commits in the range must have ICTs and
+   *                   the timestamp returned for each commit will be the ICT.
+   *                   If `false`, the file modification time will be used as the timestamp.
    */
-  def getHistory(
+  private[delta] def getHistoryImpl(
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:85]: Can we make this `end` inclusive as well (and upperBoundExclusive -> upperBoundInclusive) for consistency across different sites in this file?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 6]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
...
       start: Long,
-      endOpt: Option[Long] = None): Seq[DeltaHistory] = {
+      end: Long,
+      isRangeICT: Boolean,
+      snapshot: Snapshot): Seq[DeltaHistory] = {
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:89]: Pass `commitFileProvider` here instead of snapshot?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 7]
```diff
@@ -115,8 +117,57 @@ class DeltaHistoryManager(
...
+   * Get the commit information of the Delta table from commit `[start, end]`. If `end` is `None`,
+   * we return all commits from start to now.
+   * @param start The start of the commit range, inclusive.
+   * @param end The end of the commit range, inclusive.
+   */
+  def getHistory(
+      start: Long,
+      end: Option[Long] = None): Seq[DeltaHistory] = {
+    val currentSnapshot = deltaLog.unsafeVolatileSnapshot
+    val (snapshot, upperBoundExclusive) = end match {
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:138]: `snapshot` -> `snapshotForCommitFileProvider`
`upperBoundExclusive` -> `resolvedEnd`

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 8]
```diff
@@ -115,8 +117,57 @@ class DeltaHistoryManager(
...
+          val snapshot = deltaLog.update()
+          val endInclusive = end.getOrElse(snapshot.version).min(snapshot.version)
+          (snapshot, endInclusive + 1)
+      }
+    if (!DeltaConfigs.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetaData(snapshot.metadata)) {
+      getHistoryImpl(start, upperBoundExclusive, isRangeICT = false, snapshot)
+    } else {
+      InCommitTimestampUtils.getValidatedICTEnablementInfo(snapshot.metadata) match {
+        case Some(Commit(ictEarliest, _)) =>
+          val nonICTCommits = if (ictEarliest > start) {
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:154]: I think relying on `getHistoryImpl` to return an empty Seq when start(inclusive) > end(inclusive) is a fair assumption. It can simplify the if-else's here.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 9]
```diff
@@ -115,8 +117,57 @@ class DeltaHistoryManager(
...
+          } else {
+            Seq.empty
+          }
+          val ictCommits = if (upperBoundExclusive > ictEarliest) {
+            getHistoryImpl(
+              math.max(ictEarliest, start), upperBoundExclusive, isRangeICT = true, snapshot)
+          } else {
+            Seq.empty
+          }
+          ictCommits ++ nonICTCommits
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:166]: The ordering here is important but it requires the reader to reason about the ordering of reverse and appending lists here to prove correctness. I'm not sure if there's a more intuitive way to order them here though.

We can add some comments:
```
// Merge the two sequences, ensuring ICT commits are listed first as they are more recent,
// followed by non-ICT commits, maintaining the reverse chronological order
```

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 10]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
...
         val fs = basePath.getFileSystem(conf.value)
         versions.flatMap { commit =>
           try {
             val deltaFile = commitFileProvider.deltaFile(commit)
-            val ci = DeltaHistoryManager.getCommitInfo(logStore, deltaFile, conf.value)
-            val metadata = fs.getFileStatus(deltaFile)
-            Some(ci.withTimestamp(metadata.getModificationTime))
+            val ci = DeltaHistoryManager
+              .getCommitInfoOpt(logStore, deltaFile, conf.value)
+              .getOrElse(CommitInfo.empty(Some(commit)))
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:105]: Using an empty CommitInfo here will result in throwing [missingCommitTimestamp](https://github.com/delta-io/delta/blob/01c0ef91565d9a3e06fb9f62295f7c80fba09351/spark/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L1203) error here instead of `missingCommitInfo` error. I'm wondering if the latter would be better instead 🤔

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 11]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
...
+   *                   If `true`, all commits in the range must have ICTs and
+   *                   the timestamp returned for each commit will be the ICT.
+   *                   If `false`, the file modification time will be used as the timestamp.
    */
-  def getHistory(
+  private[delta] def getHistoryImpl(
       start: Long,
-      endOpt: Option[Long] = None): Seq[DeltaHistory] = {
+      end: Long,
+      isRangeICT: Boolean,
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:88]: Nit: A simple `useInCommitTimestamps` name would also work?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 12]
```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
...
+      math.max(snapshot.version - limit + 1, 0)
     }.getOrElse(getEarliestDeltaFile(deltaLog))
-    getHistory(listStart)
+    getHistory(listStart, end = Some(snapshot.version))
   }
 
   /**
-   * Get the commit information of the Delta table from commit `[start, end]`. If `end` is `None`,
-   * we return all commits from start to now.
+   * Get the commit information of the Delta table from commit `[start, end)`.
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:78]: Nit: Mention `in reverse chronological order`

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 13]
```diff
@@ -115,8 +117,57 @@ class DeltaHistoryManager(
...
+      info.collect()
+    } else {
+      monotonizeCommitTimestamps(info.collect())
+    }
+    monotonizedCommits.reverse
+  }
+
+  /**
+   * Get the commit information of the Delta table from commit `[start, end]`. If `end` is `None`,
+   * we return all commits from start to now.
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:130]: Nit: Mention `in reverse chronological order`

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala (hunk 5 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:85]: Done

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala (hunk 10 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:105]: `missingCommitInfo` is definitely the better error message in that case. Restructured the code to fix this.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala (hunk 7 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:138]: > upperBoundExclusive -> resolvedEnd
Done 

> snapshot -> snapshotForCommitFileProvider

Since the snapshot is used for determining the enablement status of ICT as well, I haven't changed the name.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 14]
```diff
@@ -115,8 +119,61 @@ class DeltaHistoryManager(
           }
         }.map(DeltaHistory.fromCommitInfo)
       }
-    // Spark should return the commits in increasing order as well
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:118]: Undo comment removal

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 15]
```diff
@@ -115,8 +119,61 @@ class DeltaHistoryManager(
...
+   * Get the commit information of the Delta table from commit `[start, end]` in reverse
+   * chronological order. If `end` is `None`, we return all commits from start to now.
+   * @param start The start of the commit range, inclusive.
+   * @param end The end of the commit range, inclusive.
+   */
+  def getHistory(
+      start: Long,
+      end: Option[Long] = None): Seq[DeltaHistory] = {
+    val currentSnapshot = deltaLog.unsafeVolatileSnapshot
+    val (snapshot, resolvedEnd) = end match {
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:141]: `snapshot` -> `snapshotForCommitFileProvider` since the snapshot might not necessarily be useful for other usecases.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala [hunk 16]
```diff
@@ -115,8 +119,61 @@ class DeltaHistoryManager(
...
+              useInCommitTimestamps = false,
+              commitFileProvider)
+          val ictCommits = getHistoryImpl(
+              math.max(ictEarliest, start),
+              resolvedEnd,
+              useInCommitTimestamps = true,
+              commitFileProvider)
+          // Merge the two sequences, ensuring ICT commits are listed first as they are more recent,
+          // followed by non-ICT commits, maintaining the reverse chronological order.
+          ictCommits ++ nonICTCommits
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:172]: I think it might be a problem that we don't monotonize timestamps across ICT and non-ICT commits?

Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:172]: This will only happen for tables where the file modification timestamps of non-ICT commits have been changed after the enablement of ICT (since the first ICT > last non-ICT file modification time). I think this is an acceptable tradeoff. Eventually, these non-ICT commits will expire. It is preferable to ensure that ICT timestamps are unchanging.

**Review by sumeet-db — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala (hunk 15 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala:141]: Renamed to `snapshotNewerThanResolvedEnd`.
